src/main/kotlin/org/rust/ide/refactoring/introduceParameter/impl.kt

```diff
@@ -0,0 +1,192 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+package org.rust.ide.refactoring.introduceParameter
+
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiDocumentManager
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiReference
+import com.intellij.psi.search.GlobalSearchScope
+import com.intellij.psi.search.searches.ReferencesSearch
+import com.intellij.refactoring.RefactoringBundle
+import com.intellij.refactoring.rename.inplace.MemberInplaceRenamer
+import com.intellij.refactoring.util.CommonRefactoringUtil
+import org.rust.ide.presentation.insertionSafeText
+import org.rust.ide.refactoring.*
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+import org.rust.lang.core.types.type
+import org.rust.openapiext.runWriteCommandAction
+
+fun extractExpression(editor: Editor, expr: RsExpr) {
+    try {
+        val project = expr.project
+        val enclosingMethods = findEnclosingMethods(expr)
+        when (enclosingMethods.size) {
+            0 -> {
+                val message = "No suitable method to extract parameter found."
+                val title = RefactoringBundle.message("introduce.parameter.title")
+                val helpId = "refactoring.extractParameter"
+                CommonRefactoringUtil.showErrorHint(project, editor, message, title, helpId)
+            }
+            1 -> replaceExpressionOccurrences(editor, expr, enclosingMethods.first())
+            else -> showEnclosingMethodsChooser(editor, enclosingMethods) { chosenMethod ->
+                replaceExpressionOccurrences(editor, expr, chosenMethod)
+            }
+        }
+    } catch (e: Exception) {
+        val message = e.message ?: "Unable to perform refactoring."
+        showErrorMessageForExtractParameter(expr.project, editor, message)
+    }
+}
+
+private fun replaceExpressionOccurrences(editor: Editor, expr: RsExpr, method: RsFunction) {
+    val occurrences = findOccurrences(method, expr)
+    showOccurrencesChooser(editor, expr, occurrences) { occurrencesToReplace ->
+        replaceExpression(editor, method, occurrencesToReplace)
+    }
+}
+
+private fun replaceExpression(editor: Editor, method: RsFunction, exprs: List<RsExpr>) {
+    if (exprs.isEmpty()) return
+    require(method.isWritable) { "Chosen method is not writable." }
+    val replaceForTrait = method.owner is RsAbstractableOwner.Trait || method.owner.isTraitImpl
+    val paramIntroducer = ParamIntroducer(method.project, editor)
+    paramIntroducer.replaceExpressions(method, exprs, replaceForTrait)
+}
+
+private fun findEnclosingMethods(expr: RsExpr): List<RsFunction> {
```

Comment:
`Sequence` instead of `List`
```
private fun findEnclosingFunctions(expr: RsExpr): Sequence<RsFunction> =
    expr.ancestors.filterIsInstance<RsFunction>()
```

Reply:
In this case I don't think it is useful to replace List with Sequence. We need to know the exact size of the returned collection anyway.

Reply:
OK, anyway the number of enclosing functions is small.